### DISCUSSION TITLE
Do some asynchronous processing before form submit?

### DISCUSSION THREAD

**Discussion by SpeedyCraftah** [remix-run/remix#3164] [Q&A] [answered]
This question probably does not specifically relate to Remix but I have a question regarding form submits.

I am using the form provided by Remix which provides a stateful way to handle forms without Javascript, and I have recently implemented captcha, but the problem is I want to use the invisible variant which has no button, and I would like to ideally execute the captcha on form submit, wait for the token then actually submit the form. Problem is onSubmit does not wait for any processing to finish and just submits the form, and since there is no token as it hasn't waited, the request will fail.

Any ideas?

**Comment by sergiodxa** [comment, marked as answer]
You'll have to submit the form programatically using the useSubmit hook.

```tsx
let submit = useSubmit()
async function handleSubmit(event: FormEvent<HTMLFormElement>) {
  event.preventDefault() // this will prevent Remix from submitting the form
  await doSomethingAsync() // do your async thing
  submit(event.currentTarget) // this will work as the normal Form submit but you trigger it
}
return (
  <Form ref={$form} onSubmit={handleSubmit>
    your inputs here
  </Form>
)

**Reply by SpeedyCraftah** [reply]
Never knew there was a hook for submitting. Thank you!

**Reply by SpeedyCraftah** [reply]
I just tried using it and it works... if you have nothing blocking.
I tried this as a test because my captcha code wasn't working:
```
const submit = useSubmit();

const onSubmit = (e) => {
      e.preventDefault();
      setTimeout(() => {
          submit(e.currentTarget);
      }, 1000)
}
...
<Form onSubmit={onSubmit} method="post" ...>
```

I get very weird behavior. Firstly my `useTransition` hook exits the `idle` state for a bit then comes back to it.
But the form does not submit (I can see the action did not trigger in the logs).

**Reply by sergiodxa** [reply]
Try to save `event.currentTarget` to a variable outside setTimeout

**Reply by SpeedyCraftah** [reply]
That fixed it. Thanks a bunch!
